### PR TITLE
Add cumsum pscan

### PR DESCRIPTION
**Pull request by maximzubkov**
Hello, @dvruette, and Happy New Year!

As we discussed in the [issue](https://github.com/dvruette/barrel-rec-pytorch/issues/1), there might be alternative variants of PScan. Initially, I proposed an [FFT-based implementation](https://github.com/maximzubkov/fft-scan/blob/main/src/fft_efficeint.py), but its efficiency fell short of expectations. I tried to optimize it even more and subsequently managed to come up with a simpler yet more efficient approach. The key was to replace all FFT operations with simple `cumsum`, check [this](https://github.com/maximzubkov/fft-scan/blob/main/src/cumsum.py) for more details as well as the [README](https://github.com/maximzubkov/fft-scan/blob/main/README.md) for the underlying theory.

In terms of complexity, I benchmarked the performance against PScan in your repository [by François Fleuret](https://fleuret.org/cgi-bin/gitweb/gitweb.cgi) (referred to as FF in my experiments). While both exhibit the same asymptotic complexity, FF outperforms in actual execution time. However, I observed FF's occasional instability in specific cases of matrix $A$  (this might be caused by multiple `mul_` operations resulting in float overflow). Have you encountered similar issues in your experiments? In contrast, my `cusum` implementation leverages a [trick](https://github.com/maximzubkov/fft-scan/blob/5f0828a68a4036090e8578fbedc54be8d197aace/src/cumsum.py#L17) commonly used in `softmax` computations, avoiding observed instabilities, but this needs to be checked in the real-world training setting, ofc

Benchmarking conducted with NVIDIA V100, see [Dockerfile](https://github.com/maximzubkov/fft-scan/blob/main/Dockerfile) for the environment details, and [benchmark.py](https://github.com/maximzubkov/fft-scan/blob/main/benchmark.py). I didn't include `pscan_fft.py` in the benchmarking since the results are much worse than `ff` and `cumsum`, in line with what you observed running the training.

f
fb

### CODE REVIEW

**Comment by maximzubkov** [comment]
And both perform almost the same on the `wiki` dataset
Screenshot 2024-01-03 at 16 58 07

**Comment by dvruette** [comment]
Very cool, thanks for testing all this! Yes, I think I'm vaguely aware that the reference implementation can have numerical instabilities, but it doesn't seem to be a problem in practice.

Quite unfortunate that the cumsum approach turns out to be slower, I think Francois mentioned on Twitter that he initially was also using cumsum but struggled to get an efficient backward pass. The thing I'm looking for first and foremost right now is faster training times, so I'm down to merging this, but I'm probably not going to use it until we find something that's faster than what I'm using now.
